js/admin.js

```diff
@@ -32,75 +39,150 @@ jQuery( document ).ready(
+			};
+		}
+
+		// Create the selectmenu widget only if the field is present.
+		if ( flagListExist ) {
+			var selectmenuFlagList = $( '#flag_list' ).selectmenu( Object.assign( {}, selectmenuOptions, selectmenuFlagListCallbacks ) ).selectmenu( 'instance' );
+			// Overrides each item in the jQuery UI selectmenu list by injecting flag image.
+			selectmenuFlagList._renderItem = selectmenuRenderItem;
+			// Override the selected item rendering for jQuery UI 1.12
+			if ( isJqueryUImin112 ) {
+				selectmenuFlagList._renderButtonItem = selectmenuRenderButtonItem;
+				selectmenuFlagList.refresh(); // Need to refresh to take in account the button item rendering method after the selectmenu widget instanciaion.
 			}
-		);
-		// Create the jQuery UI selectmenu widget
-		$( '#lang_list' ).selectmenu( { width: '97%' } ); // jQuery UI selectmenu widget substract 2% and we need 95% for the width matches to the other fields width.
-		// However for the wizard we need a 100% width.
-		if( $( '#lang_list' ).closest( '.pll-wizard-content' ).length > 0 ) {
-			$( '#lang_list' ).selectmenu( 'option', 'width', '102%' );
 		}
-		// Overrides each element in the jQuery UI selectmenu list by injecting flag image.
-		$( '#lang_list' ).selectmenu( 'instance' )._renderItem = selectmenuRenderItem;
+		// Language choice in predefined languages in Polylang Languages settings page and wizard.
+		// Overrides the predefined language dropdown list with our customized jQuery ui selectmenu widget.
 
-		// Languages form
-		// Fills the fields based on the language dropdown list choice
-		$( '#add-lang #lang_list' ).on(
-			'selectmenuchange',
-			function() {
-				var value = $( this ).val().split( ':' );
-				var selected = $( "option:selected", this ).text().split( ' - ' );
-				$( '#lang_slug' ).val( value[0] );
-				$( '#lang_locale' ).val( value[1] );
-				$( 'input[name="rtl"]' ).val( [value[2]] );
-				$( '#lang_name' ).val( selected[0] );
+		// Callback when selectmenu widget change event is triggered.
+		var changeCallback = function( event, ui ) {
+			var value = $( event.target ).val().split( ':' );
+			var selected = $( "option:selected", event.target ).text().split( ' - ' );
+			$( '#lang_slug' ).val( value[0] );
+			$( '#lang_locale' ).val( value[1] );
+			$( 'input[name="rtl"]' ).val( [value[2]] );
+			$( '#lang_name' ).val( selected[0] );
+
+			// Refresh the flag field only if it's present.
+			if ( flagListExist ) {
 				$( '#flag_list').val( value[3] );
 
-				// Refresh the jQuery UI selectmenu flags list.
-				$( '#flag_list' ).selectmenu( 'refresh' ).trigger( 'selectmenuselect' );
+				// Refresh the jQuery UI selectmenu flag list.
+				if ( isJqueryUImin112 ) {
+					// Just a refresh of the menu is needed with jQuery UI 1.12 because _renderButtonItem is triggered and then inject correctly the flag.
+					selectmenuFlagList.refresh();
+				} else {
+					selectmenuRefreshButtonText( $( '#flag_list').selectmenu( 'refresh' ) );
+				}
 			}
```

Comment:
You can create a custom event with a combination of jQuery's `.on()` and `.trigger()` method. You could then trigger a 'languagesChanged' event here and handle it with a callback from `selectmenuFlagList`.

That way, `selectmenuLangList` (which is used in both Languages settings page and Wizard's Languages step) doesn't need to know about `selectmenuFlagList` (which is only used in Languages settings page).

Reply:
🤔 Hum I'm not sure we don't need a reference to the flag list widget. Because .on() and .trigger() methods need to be attached to an element.
See https://api.jquery.com/on/
and https://api.jquery.com/trigger/#trigger-eventType-extraParameters

It's the same with jQuery UI widget factory methods - ._on() and ._trigger - used to construct selemenu widget and which use internally the jQuery .on() and .trigger().

However it's interesting because we can use the same opencallbacks function and process in selectmenuFlagList widget. That's the right place to do that.

Reply:
So, rather than using the jQuery `.on()` method and to be consistent with the other callbacks I used, I preferred to declare the `change` event directly as a callback when we create the selectmenu flags list widget.
As the code is exactly the same as the open events, I renamed the `openCallback` to changeOpenCallback.

Then because the condition of jQuery UI version to use is already tested when we build the callbacks configuration for the selectmenu widget, it isn't necessary anymore to test the condition to refresh the selectmenu flags list widget.

The selectmenu flags list widget is now simply refresh by triggering its `change` event.